ipld/get_namespaced_shares.go

```diff
@@ -0,0 +1,186 @@
+	mu      sync.Mutex
+	lowest  int
+	highest int
+}
+
+func (b *fetchedBounds) Update(index int) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if index <= b.lowest {
+		b.lowest = index
+	}
+	// not an `else if` because an element can be both the
+	// lower and higher bound
+	if index >= b.highest {
+		b.highest = index
+	}
+}
+
+// getLeavesByNamespace returns all the leaves from the given root with the given namespace.ID.
+// If nothing is found, it returns both data and error as nil.
+// The following implementation is based on `GetShares`.
+func getLeavesByNamespace(
+	ctx context.Context,
+	bGetter blockservice.BlockGetter,
+	root cid.Cid,
+	nID namespace.ID,
+	maxShares int,
+) ([]ipld.Node, error) {
+	err := SanityCheckNID(nID)
+	if err != nil {
+		return nil, err
+	}
+
+	// we don't know where in the tree the leaves in the namespace are,
+	// so we keep track of the bounds to return the correct slice
+	// maxShares acts as a sentinel to know if we find any leaves
+	bounds := fetchedBounds{sync.Mutex{}, maxShares, 0}
+
+	// buffer the jobs to avoid blocking, we only need as many
+	// queued as the number of shares in the second-to-last layer
+	jobs := make(chan *job, (maxShares+1)/2)
+	jobs <- &job{id: root}
+
+	// the wg counter cannot be pre-allocated either, it is incremented with each job
+	wg := wrappedWaitGroup{sync.WaitGroup{}, sync.Mutex{}, 0}
+	wg.Add(1)
+
+	// we overallocate space for leaves since we do not know how many we will find
+	// on the level above, the length of the Row is passed in as maxShares
+	leaves := make([]ipld.Node, maxShares)
+	leafLk := sync.Mutex{}
+
+	// if the WaitGroup counter is above 0, we are still walking the tree
+	for wg.IsWorking() {
+		select {
+		case j := <-jobs:
+			namespacePool.Submit(func() {
+				defer wg.Done()
+
+				rootH := plugin.NamespacedSha256FromCID(j.id)
+				if nID.Less(nmt.MinNamespace(rootH, nID.Size())) || !nID.LessOrEqual(nmt.MaxNamespace(rootH, nID.Size())) {
+					return
+				}
+
+				nd, err := plugin.GetNode(ctx, bGetter, j.id)
+				if err != nil {
+					return
+				}
+
+				links := nd.Links()
+				linkCount := len(links)
+				// wg counter needs to be incremented **before** adding new jobs
+				if linkCount > 1 {
+					wg.Add(linkCount)
+				} else if linkCount == 1 {
+					// successfully fetched a leaf belonging to the namespace
+					leafLk.Lock()
+					leaves[j.pos] = nd
+					leafLk.Unlock()
+					// we found a leaf, so we update the bounds
+					bounds.Update(j.pos)
+					return
+				}
+
+				// this node has links in the namespace, so keep walking
+				for i, lnk := range links {
+					select {
+					case jobs <- &job{
+						id: lnk.Cid,
+						// position represents the index in a flattened binary tree,
+						// so we can return a slice of leaves in order
+						pos: j.pos*2 + i,
+					}:
+					case <-ctx.Done():
+						return
+					}
+				}
+			})
+		case <-ctx.Done():
+			return nil, nil
+		// default case is needed to prevent blocking
+		default:
```

Comment:
Ok yup, i'll try again - couldn't figure out a way to get it working yet without the default case. But that was before I was able to check IsWorking

Reply:
@Wondertan Ok, I definitely made some progress with it if you check the latest commit. We now always block on jobs, and close the channel when the wg counter reaches 0. This means we only check if we are done once per job.

One issue remains to avoid a potential race condition: What is the best pattern to close a channel atomically? Is a mutex okay here?

Reply:
I think the current logic in wrappedWaitGroup closes the channel well enough and won't cause any issues with how it is ___currently___ used. If wrappedWaitGroup is incorrectly used, then yes, it may cause issues, which can be prevented by additional safety measures, but as long wrappedWaitGroup is not abstracted and private, I think it is fine to keep things the way they are currently.